syntax = "proto3";

package flyteidl2.plugins;

option go_package = "github.com/flyteorg/flyte/v2/gen/go/flyteidl2/plugins";

// Custom proto for torch elastic config for distributed training using
// https://github.com/kubeflow/trainer/blob/e31d11faa9f6ce5111b60c01079d39295589e0ef/pkg/apis/kubeflow.org/v1/pytorch_types.go#L98
message ElasticConfig {
  string rdzv_backend = 1;
  int32 min_replicas = 2;
  int32 max_replicas = 3;
  int32 nproc_per_node = 4;
  int32 max_restarts = 5;
}

// Custom proto for plugin that enables distributed training using https://github.com/kubeflow/trainer
message DistributedPyTorchTrainingTask {
  // number of worker replicas spawned in the cluster for this job
  int32 workers = 1;

  // config for an elastic pytorch job
  ElasticConfig elastic_config = 2;
}
